common/src/main/java/org/apache/cassandra/sidecar/common/CassandraAdapterDelegate.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.sidecar.common;
+
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import com.google.common.base.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.Host;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.exceptions.NoHostAvailableException;
+
+
+/**
+ * Since it's possible for the version of Cassandra to change under us, we need this delegate to wrap the functionality
+ * of the underlying Cassandra adapter.  If a server reboots, we can swap out the right Adapter when the driver
+ * reconnects.
+ *
+ * This delegate *MUST* checkSession() before every call, because:
+ *
+ * 1. The session lazily connects
+ * 2. We might need to swap out the adapter if the version has changed
+ *
+ */
+public class CassandraAdapterDelegate implements ICassandraAdapter, Host.StateListener
+{
+    private final CQLSession cqlSession;
+    private final CassandraVersionProvider versionProvider;
+    private Session session;
+    private CassandraVersion currentVersion;
+    private ICassandraAdapter adapter;
+    private Boolean isUp = false;
+    private final int refreshRate;
+
+    private static final Logger logger = LoggerFactory.getLogger(CassandraAdapterDelegate.class);
+    private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+    private boolean registered = false;
+
+    public CassandraAdapterDelegate(CassandraVersionProvider provider, CQLSession cqlSession)
+    {
+        this(provider, cqlSession, 5000);
+    }
+
+    public CassandraAdapterDelegate(CassandraVersionProvider provider, CQLSession cqlSession, int refreshRate)
+    {
+        this.cqlSession = cqlSession;
+        this.versionProvider = provider;
+        this.refreshRate = refreshRate;
+    }
+
+    public synchronized void start()
+    {
+        logger.info("Starting health check");
+        // TODO: maybe unhardcode
```

Comment:
Do you want to address this before merging?

Reply:
Ah yes, I originally had the time hard coded, removed it, then forgot to remove the comment.